### DISCUSSION TITLE
Bash shell integration makes script execution slow

### DISCUSSION THREAD

**Discussion by ZoltanLajosKis** [ghostty-org/ghostty#3724] [Issue Triage] [answered]
Script execution with shell-integration takes a lot longer than with `shell-integration = none`.

Executing `time for i in {0..100000}; do :; done` takes around **150 ms** when disabled, and takes over **4 seconds** when not disabled.

Executing `time for i in {0..100000}; do echo $i; done` takes around **450 ms** when disabled, and takes over **6 seconds** when not disabled.

Tested on Ubuntu 22.04.5 with X11, running Ghostty 1.0.0 built with zig-0.13.0 (--release=fast), default config.

**Comment by jparise** [comment]
I have an upstream `bash-preexec` change under consideration that should improve this by removing per-prompt `history | sed` command overhead: https://github.com/rcaloras/bash-preexec/pull/167#issuecomment-2587093723

**Reply by jparise** [reply]
I merged a version of this as #5141. It's available in Ghostty tip if you'd like to give that a try before the next release.

**Reply by jparise** [reply]
@ZoltanLajosKis Ghostty 1.1.0 includes these changes. Has it improved your experience?

**Reply by ZoltanLajosKis** [reply]
@jparise I still see the same execution times with 1.1.0.

**Reply by jparise** [reply, marked as answer]
The remaining overhead (which your example directly demonstrates) is attributable to bash-preexec's `DEBUG` trap, which unfortunately runs much more often than we actually need (once before each interactive command). This has been an issue for a while: https://github.com/Bash-it/bash-it/issues/1943

For the moment, there isn't an easy solution to removing that overhead. Newer versions of bash (5.3+) may offer alternatives via `PS0`, but exploring that will require a round of focused work and testing, so I can't offer a timeline.

**Reply by ZoltanLajosKis** [reply]
Thanks for looking into this.

**Comment by jparise** [comment]
This is fixed by #10609 (planned for the 1.3 release) for bash 4.4+.

```
$ time for i in {0..100000}; do :; done

real    0m0.133s
user    0m0.128s
sys     0m0.004s
```
